### DISCUSSION TITLE
[Request] Replace focus on semi-auto-properties with focus on property-scoped fields

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6632] [General]
There was a lot of talk for C#11 about semi-auto-properties, but it was scrapped for 11.

I want to propose instead focusing on property-scoped fields. They would work like:
``` csharp
public class C
{

    public string ValueTypeMember
    {
        string? backingField;
        get => backingField ??= CalculateStringValue();
        set => value = backingField;
    }

    public int ReferenceTypeMember
    {
        int? backingField;
        get => backingField ??= CalculateIntValue();
        set => value = backingField;
    }


    public void SomeMethod()
    {
        _ = backingField; // error: neither backingField's are in scope
    }
}
```

I am really interested in the functionality of semi-auto-properties, but this proposal would cover 100% of s-a-p scenarios, and add additional useful functionality. 

For example, the nullability of the backing field, For reference types with NRT enabled, it would be iffy to make the backing field nullable where the property type isn't. For value types, that's a definite no-go.

**Comment by RenderMichael** [comment]
Just found that this proposal has an associated issue https://github.com/dotnet/csharplang/issues/133

**Reply by CyrusNajmabadi** [reply]
Ldm already decided that we'd rather go with three simpler approach that covers the majority of scenarios.

**Reply by jnm2** [reply]
With the simpler approach first, but open to #133 potentially soon after, right? I've brought up wanting #133 almost right away because of this nullable value type situation.

**Comment by jnm2** [comment]
#140 is still a good initial focus because of how much less code it is for many more scenarios. I agree with this lazy value type pain and am hoping #133 happens soon afterwards since #140 doesn't improve that situation.

Most of the time, properties like this don't have setters. This is what I would prefer:

```cs
    // One-liner. #140
    public string ReferenceTypeMember => field ??= CalculateStringValue();

    // One-liner with a manually-declared field. #133
    public int ValueTypeMember { int? field; get => field ??= CalculateIntValue(); }
```
